### ISSUE TITLE
Terraform refresh is not working with Terraform Cloud remote backend

### ISSUE THREAD

**Issue by CarpathianUA** [hashicorp/terraform#23247]
### Terraform Version
Terraform v0.12.10
+ provider.aws v2.33.0

Hey guys, I'm using Terraform Cloud as a `remote` backend.
For example, I've changed output and nothing else and need it to be updated in a state. 
`terraform refresh`
```
Error: error starting operation:

The "remote" backend does not support the "OperationTypeRefresh" operation.
```
How can I achieve this with TF Cloud?

**Comment by CarpathianUA** [comment]
Any thoughts on that?

**Comment by CarpathianUA** [comment]
Any updates?

**Comment by hspak** [comment]
I am also interested in this.

Is there a current workaround to refreshing the output aside from pushing a no-op change to apply?

**Comment by sandangel** [comment]
I have to change backend from remote to local, run refresh, and change back to remote

**Comment by hamishforbes** [comment]
Just run into this myself.  
Trying to migrate a monolithic TF config to multiple TF cloud workspaces with remote state, which is very annoying when you have to make dummy changes to every module to update outputs.

Any chance of refresh being supported on the TF cloud remote backend?

**Comment by CarpathianUA** [comment]
Really strange for me that a TF Cloud as a paid product doesn't have this obvious and simple fucntionality...
Hiashicorp guys, any updates? We realy need this!

**Comment by sethbacon** [comment]
Even a button in the TF Cloud interface, similar to the Queue Plan button, but for Refresh. OR a check box on the settings for the workspace. Anything would help here...

**Comment by gianarb** [comment]
same issue here, my state is broken and I am not sure about how to get it fixed

**Comment by MarcPearce** [comment]
whats the status of this ?

**Comment by CarpathianUA** [comment]
Looks like Hashicorp totally ignores the issue

**Comment by jgitlin-p21** [comment]
Had the same situation here: added an `output` but made no other changes and was not able to use `terraform refresh` as suggested by `terraform output`:

> The state file either has no outputs defined, or all the defined
> outputs are empty. Please define an output in your configuration
> with the `output` keyword and run `terraform refresh` for it to
> become available. If you are using interpolation, please verify
> the interpolated value is not empty. You can use the 
> `terraform console` command to assist.

I had to do what @sandangel suggested, and did that as follows:

```
# Optional: Lock the workspace using Settings > Locking from within app.terraform.io
terraform state pull > terraform.tfstate
# edit config.tf and comment out the `backend "remote"` block
terraform init
terraform refresh
terraform output
# edit config.tf and uncomment out the `backend "remote"` block
terraform init
```

Not fun, but it is a viable workaround until this bug is fixed.

**Comment by sethbacon** [comment]
Related: https://github.com/hashicorp/terraform/issues/22607

**Comment by dgreenfield0** [comment]
As a temporary workaround, I’ve added an extra TAG to my resources, which can be bumped/updated as needed, allowing an “update” to cause a refresh without affecting the actual resource(s).

```
tags = {
    Refresh  = var.tag_refresh_value
}
```

**Comment by maetthu** [comment]
Another workaround without interfering with real resources

```hcl
resource "random_pet" "tfc_refresh" {
  keepers = {
    refresh : 1
  }
}
```

I just increment the counter when I want to force refreshing the state.

**Comment by Myatu** [comment]
Nice one @maetthu . I've modified it slightly so it just picks up any change to the `outputs.tf` file in my case:

```
resource "random_uuid" "tfc_output_refresh" {
  keepers = {
    refresh : filesha256("${path.module}/outputs.tf")
  }
}
```

**Comment by hamishforbes** [comment]
I have, once again, been bitten by this issue.

This time I am trying to upgrade a Terraform Cloud workspace from 0.12.x to 0.13.x
In some cases the apply step when running remotely is failing to initialise providers (not sure why this works fine in planning and locally but go figure...)

Workaround is to again disable remote execution, run terraform refresh and re-enable remote execution

**Comment by apparentlymart** [comment]
Hi all! Sorry for the slow response here; this repository is typically monitored only by the developers of Terraform CLI, whereas remote operations is maintained by the Terraform Cloud teams who don't tend to see bug reports and feature requests logged here. (If you want to make requests for Terraform Cloud in particular, you can contact Terraform Cloud support at [tf-cloud-support@hashicorp.com](mailto:tf-cloud-support@hashicorp.com).)

One reason why the team hasn't prioritized making `terraform refresh` work with remote operations is that the Terraform Core team is working towards obsoleting the `terraform refresh` command by gradually incorporating all of its use-cases into the main `terraform plan`/`terraform apply` workflow. Recent Terraform CLI releases have included changes in that vein, including now treating changes to root module outputs as something that can be planned and applied even if there aren't any resource changes to perform. That particular change arrived initially in Terraform v0.13 and saw some further improvements in both Terraform v0.14 and v0.15.

The discussion here seems to mainly focus on updating outputs, in which case hopefully the changes in Terraform v0.13 and later have resolved that. There is still a goal for the remote backend to support all of the operations that the local backend does when it's in remote operations mode, so I'm going to leave this open to continue to represent that `terraform refresh` isn't supported there, though it remains to be seen whether it'll ultimately be resolved by the Terraform Cloud teams implementing remote refresh or by the `terraform refresh` command becoming obsolete (by incorporating all of its use-cases into `terraform plan`/`terraform apply`) first.

Hopefully the improvements to `terraform plan` in Terraform v0.13 will address the specific friction you all were discussing here, at least.

**Comment by eriksw** [comment]
@apparentlymart One case where we desperately need `refresh` with terraform cloud is when upgrading a **ton** of workspaces from 0.12 to 0.14 and the requirement that the state be rewritten by 0.13 before going to 0.14.

**Comment by eriksw** [comment]
@apparentlymart As a follow-up, consider this a request that you do not obsolete/remove `terraform refresh` until the requirement to rewrite state using intermediate versions is gone.

If you're going to require that the state be written by n-1 minor version, there _needs_ to be a way to force that state rewriting to happen even when there are no resource or output changes.

**Comment by apparentlymart** [comment]
Hi @eriksw,

It sounds like you've encountered a bug in the v0.12 to v0.13 upgrade process where running `terraform apply` wasn't sufficient to complete the state part of the upgrade. If you still have some details about that then I'd love to hear more about it in a new issue, where we might be able to address it in a new v0.13 point release. (Issues that block folks from upgrading are one situation where we can make exceptions to our usual backport policy, because of course you can't benefit from a fix in a newer release if you can't get upgraded.)

The instruction to run `terraform apply` to update the state as part of the v0.13 upgrade was something specific to that particular upgrade, so future changes related to `terraform refresh` wouldn't either help or hinder that: even if a newer version of Terraform CLI added support for `terraform refresh` in Terraform Cloud (which would first require Terraform Cloud to support refreshing, of course), that wouldn't help with upgrading to v0.13 because that version still wouldn't have support for remotely executing that command.

---

Incidentally, if you are running Terraform CLI locally with Terraform Cloud credentials that grant you access to the state of the target workspace, you can ask Terraform CLI to skip using remote operations just for one command by setting the environment variable `TF_FORCE_LOCAL_BACKEND=1`. This is actually an implementation detail of how Terraform Cloud knows not to recursively start another level of remote operation when it's running Terraform remotely, and so not something guaranteed to be available in all future versions of Terraform CLI, but since we're talking specifically about the Terraform v0.13 upgrade it's safe to say that this environment variable is available in all of the v0.13 releases, allowing for the following workaround to do a one-shot local refresh:

```
TF_FORCE_LOCAL_BACKEND=1 terraform refresh
```

The implementation of that environment variable prevents the remote backend code from reaching the error message shown in the original issue writeup:

https://github.com/hashicorp/terraform/blob/afbde358fad851b80e7fecab7dbe4833f49092ae/backend/remote/backend.go#L708-L726

**Comment by eriksw** [comment]
@apparentlymart I worked around my specific case with null no-op exec's, but to elaborate on situation a little more:

I don't operate with having full access to our cloud resources as myself, the creds with that access are _exclusively_ in write-only secrets in Terraform Cloud (with some break-glass options that are _extremely_ painful to activate).

With local apply and refresh out of the question due to lack of access, the experience of upgrading from 0.12 to 0.14 was pretty miserable because I had to use null exec dummy resource changes to force an in-the-cloud apply.

I appreciate the pointer to the local backend forcing, but the only thing that would've helped was a way to force the refresh to happen in the cloud (where it has access to the secrets).

I'm glad to hear the 0.13 change was a one-off; I do hope that the "local plan/apply/refresh is _not possible_" use case will be given thought in the future if anything similar comes along.

**Comment by phillhocking** [comment]
> Hi all! Sorry for the slow response here; this repository is typically monitored only by the developers of Terraform CLI, whereas remote operations is maintained by the Terraform Cloud teams who don't tend to see bug reports and feature requests logged here. (If you want to make requests for Terraform Cloud in particular, you can contact Terraform Cloud support at [tf-cloud-support@hashicorp.com](mailto:tf-cloud-support@hashicorp.com).)
> 
> One reason why the team hasn't prioritized making `terraform refresh` work with remote operations is that the Terraform Core team is working towards obsoleting the `terraform refresh` command by gradually incorporating all of its use-cases into the main `terraform plan`/`terraform apply` workflow. Recent Terraform CLI releases have included changes in that vein, including now treating changes to root module outputs as something that can be planned and applied even if there aren't any resource changes to perform. That particular change arrived initially in Terraform v0.13 and saw some further improvements in both Terraform v0.14 and v0.15.
> 
> The discussion here seems to mainly focus on updating outputs, in which case hopefully the changes in Terraform v0.13 and later have resolved that. There is still a goal for the remote backend to support all of the operations that the local backend does when it's in remote operations mode, so I'm going to leave this open to continue to represent that `terraform refresh` isn't supported there, though it remains to be seen whether it'll ultimately be resolved by the Terraform Cloud teams implementing remote refresh or by the `terraform refresh` command becoming obsolete (by incorporating all of its use-cases into `terraform plan`/`terraform apply`) first.
> 
> Hopefully the improvements to `terraform plan` in Terraform v0.13 will address the specific friction you all were discussing here, at least.

Well here we are at 1.0.0, and I am still getting bit by this. What is the status on this request and how can we ensure that refresh-only plans can be executed? 

![image](https://user-images.githubusercontent.com/4030898/121926290-a9ed8680-ccf2-11eb-8f8b-b15297a5d3ea.png)

**Comment by phillhocking** [comment]
I received a response from [tf-cloud-support@hashicorp.com](mailto: tf-cloud-support@hashicorp.com) that I thought might be helpful for folks: 

> Hello Phillip,
> 
> Thank you for contacting HashiCorp Support.
> 
> The Refresh-only run option has been added recently to Terraform Cloud and can be initiated either via CLI or API (and the UI coming soon).
> 
> For example, to make a refresh-only run for a particular workspace through the TF Cloud Run API endpoint you can use the following request:
> 
> ```
> curl \
>   --header "Authorization: Bearer $TOKEN" \
>   --header "Content-Type: application/vnd.api+json" \
>   --request POST \
>   --data @payload.json \
>   https://app.terraform.io/api/v2/runs
> ```
> and adjust the id of the workspace to match yours in the payload.json
> 
> ```
> {
>   "data": {
>     "attributes": {
>       "refresh-only": true
>     },
>     "type":"runs",
>     "relationships": {
>       "workspace": {
>         "data": {
>           "type": "workspaces",
>           "id": "ws-xxxxxxxxxxxxx"
>         }
>       }
>     }
>   }
> }
> ```
> Note that in TF Cloud the refresh-only run is supported for Terraform version 0.15.4+.
> 
> I hope that the provided information proves helpful.
> 
> Thanks.
> 
> Regards,
> Atanas
> HashiCorp Support

However, when I attempted using `terraform` via cli to do this, here is my outcome: 

```
When attempting to do this via the cli this is my result: 

PS C:\Users\phillip.hocking\Workspace\aws-ghost> terraform -version
Terraform v1.0.0
on windows_amd64
+ provider registry.terraform.io/hashicorp/aws v3.44.0
+ provider registry.terraform.io/hashicorp/null v3.1.0
PS C:\Users\phillip.hocking\Workspace\aws-ghost> terraform plan -refresh-only
Running plan in the remote backend. Output will stream here. Pressing Ctrl-C
will stop streaming the logs, but will not stop the plan running remotely.
Preparing the remote plan...
To view this run in a browser, visit:
https://app.terraform.io/app/phillhocking/aws-ghost-dev/runs/run-rDWi3e5xtqqkDqMj
Waiting for the plan to start...
Terraform v1.0.0
on linux_amd64
Configuring remote state backend...
Initializing Terraform configuration...
null_resource.terraform-github-actions: Refreshing state... [id=5045453508422674158]
module.ghost-lightsail.aws_route53_zone.green_no_www: Refreshing state... [id=Z021635518BR112CMNW3M]
aws_security_group.ghost-lightsail: Refreshing state... [id=sg-0f2532e6ed7370c25]
module.ghost-lightsail.aws_lightsail_instance.ghost-dev: Refreshing state... [id=ghost-lightsail-dev]
module.ghost-lightsail.aws_cloudfront_distribution.ghost-lightsail: Refreshing state... [id=E22VB91CZF3ST3]
module.ghost-lightsail.aws_lightsail_static_ip.dev_static_ip: Refreshing state... [id=dev_static_ip]
module.ghost-lightsail.aws_route53_record.green_no_www: Refreshing state... [id=Z021635518BR112CMNW3M_phillhocking.com._A]
module.ghost-lightsail.aws_lightsail_static_ip_attachment.static_ip_attach: Refreshing state... [id=dev_static_ip]
Note: Objects have changed outside of Terraform
Terraform detected the following changes made outside of Terraform since the
last "terraform apply":
  # module.ghost-lightsail.aws_lightsail_instance.ghost-dev has been changed
  ~ resource "aws_lightsail_instance" "ghost-dev" {
        id                 = "ghost-lightsail-dev"
      ~ is_static_ip       = false -> true
        name               = "ghost-lightsail-dev"
      ~ public_ip_address  = "*snip*" -> "*snip*"
        tags               = {
            "env" = "dev"
        }
        # (13 unchanged attributes hidden)
    }
This is a refresh-only plan, so Terraform will not take any actions to undo
these. If you were expecting these changes then you can apply this plan to
record the updated values in the Terraform state without changing any remote
objects.
```
A run/plan was not triggered in Terraform Cloud, when instantiating a run via the ui here is my outcome: 

![image](https://user-images.githubusercontent.com/4030898/122052833-f8eef680-cd9a-11eb-8a07-650e55423e6f.png)

I personally don't want to have to `curl` to the api manually to queue this change. Can someone shed light on why there's such a disconnect between the `terraform` and Terraform Cloud team? Break down silos, guys!

**Comment by apparentlymart** [comment]
Hi all,

Indeed, the changes I was describing earlier to replace `terraform refresh` with new options to `terraform plan` and `terraform apply` are now available as of Terraform v0.15.4, which means it's also in Terraform v1.0.

It seems like the support team advised @phillhocking to use the API-driven version of this facility to trigger a "refresh-only" run from Terraform Cloud, which is certainly one way to do it. Another way to do it would be to run a command like this:

```
terraform apply -refresh-only
```

If you have the remote backend enabled then Terraform CLI v1.0 should respond to that command by calling the very same API that the support team suggested. As usual with `terraform apply`, Terraform will then report what it intends to do and ask you to confirm it by typing `yes`. The difference compared to a normal plan is that Terraform won't try to compare current resource configurations with the freshly-updated state in order to propose any additional changes.

It looks like you ran `terraform plan -refresh-only`, which is also a valid command to run but it creates a _speculative plan_ -- one that just shows what the effect would be, without any intent of applying it -- and so in that case Terraform showed what it detected but didn't take any action. That seems to match the behavior I would expect for `terraform plan`, whether or not using Terraform Cloud.

In my previous comment I mentioned that it was not yet decided how the Cloud team was going to respond to the replacement of `terraform refresh` by `terraform apply -refresh-only`, but I can now confirm that refresh-only plans are the intended replacement for `terraform refresh` in both Terraform CLI and in Terraform Cloud. There's some additional context about this in [the `terraform refresh` documentation](https://www.terraform.io/docs/cli/commands/refresh.html).

We have no plans to remove `terraform refresh` from Terraform CLI for the forseeable future, because we know that it's in relatively wide use in various automation and there isn't any justification to break that automation. However, since it never existed at all for Terraform Cloud, the tradeoff for implementing it there is significantly different, and so the Terraform Cloud team has instead focused on implementing its replacement, refresh-only plans. They've also added support for the other new planning option `-replace=...`, and so you can now remotely create and apply plans to replace particular objects if they are degraded or unfit in some way, a missing feature that `terraform taint` was previously often used as a workaround for.

The future direction for Terraform CLI is for all significant use-cases to eventually be met by `terraform plan` and `terraform apply`, either by new command line options or by new configuration language features depending on the use-case, and so support for those new mechanism should also typically be available in Terraform Cloud remote operations as soon as available in Terraform CLI. Neither team plans to do much more deep investment in the various ancillary commands that essentially subvert the plan and apply process by taking actions "behind its back".

Given that, Terraform Cloud in combination with Terraform v1.0 should now meet the use-case this issue represents -- being able to incorporate changes from outside of Terraform into the Terraform state without taking other actions -- and so I'm going to close this. As usual, if you have specific questions or feedback about using this within Terraform Cloud, or other related questions such as workflow for upgrading to new versions of Terraform CLI within Cloud, then you can contact the Terraform Cloud support team at tf-cloud-support@hashicorp.com . If you have feedback about the _CLI_ portions of this then please do open a new feature request issue to discuss those! Thanks for raising this!
